Scripts/Runtime/Jobs/Collections/CollectionAccessController.cs

```diff
@@ -0,0 +1,226 @@
+            m_ExclusiveWriteDependency = m_SharedWriteDependency = m_SharedReadDependency = initialDependency;
+
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+            m_AcquireCallerInfo = string.Empty;
+            m_ReleaseCallerInfo = string.Empty;
+            m_LastHandleAcquired = default;
+#endif
+        }
+
+        /// <summary>
+        /// Returns a <see cref="JobHandle"/> to schedule a job off of based on the desired <see cref="AccessType"/>.
+        /// </summary>
+        /// <param name="accessType">The type of access to schedule the job off of.</param>
+        /// <returns>The <see cref="JobHandle"/> to wait upon.</returns>
+        public JobHandle Acquire(AccessType accessType)
+        {
+            Debug.Assert(!IsDisposed);
+            ValidateAcquireState();
+
+            JobHandle acquiredHandle = default;
+            switch (accessType)
+            {
+                case AccessType.ForDisposal:
+                    m_State = AcquisitionState.Disposing;
+                    acquiredHandle = m_ExclusiveWriteDependency;
+                    break;
+                case AccessType.ExclusiveWrite:
+                    m_State = AcquisitionState.ExclusiveWrite;
+                    acquiredHandle = m_ExclusiveWriteDependency;
+                    break;
+                case AccessType.SharedWrite:
+                    m_State = AcquisitionState.SharedWrite;
+                    acquiredHandle = m_SharedWriteDependency;
+                    break;
+                case AccessType.SharedRead:
+                    m_State = AcquisitionState.SharedRead;
+                    acquiredHandle = m_SharedReadDependency;
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(accessType), accessType, $"Tried to acquire with {nameof(AccessType)} of {accessType} but no code path satisfies!");
+            }
+
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+            m_LastHandleAcquired = acquiredHandle;
+#endif
+
+            return acquiredHandle;
+        }
+
+        /// <summary>
+        /// Allows the <see cref="CollectionAccessController{TContext,TKey}"/> to be aware of the work that you are doing
+        /// for a specific <see cref="AccessType"/>. You must call <see cref="Release"/> after any call to <see cref="Acquire"/>
+        /// before you call <see cref="Acquire"/> again.
+        /// </summary>
+        /// <param name="releaseAccessDependency">The <see cref="JobHandle"/> to the job that is doing the reading or
+        /// writing from/to the underlying collection this <see cref="CollectionAccessController{TContext,TKey}"/>
+        /// is gating access to.</param>
+        public void Release(JobHandle releaseAccessDependency)
+        {
+            Debug.Assert(!IsDisposed);
+            ValidateReleaseState(releaseAccessDependency);
+
+            switch (m_State)
+            {
+                case AcquisitionState.Disposing:
```

Comment:
Don't you want to throw an exception if you're trying to release on a disposing state?

Reply:
We do, down in `ValidateReleaseState`

Reply:
Yea but that's only with the collection check enabled.

Reply:
Not anymore. See the switch statement in the updated code.

Reply:
lol, ok. Was responding to your first comment!